tests/kuberay_utils/utils.py

```diff
@@ -123,71 +122,77 @@ def wait_for_condition(
-        # The default container restartPolicy of a Pod is `Always`. Hence, when GCS server is killed,
-        # the head pod will restart the old one rather than create a new one.
-        if new_head_pod_name != old_head_pod_name:
-            logger.info(f'If GCS server is killed, the head pod will restart the old one rather than create a new one.' +
-                f' new_head_pod_name: {new_head_pod_name}, old_head_pod_name: {old_head_pod_name}')
-            # TODO (kevin85421): We should `return False` here, but currently ray:nightly has a high possibility to create
-            #                    a new head pod instead of restarting the old one.
-
-        # When GCS server is killed, it takes nearly 1 min to kill the head pod. In the minute, the head
-        # pod will still be in 'Running' and 'Ready'. Hence, we need to check `restart_count`.
-        else:
-            # TODO (kevin85421): We should remove `else` in the future. Currently, ray:nightly has a high possibility to
-            #                    create a new head pod instead of restarting the old one. The new pod's `restart_count`
-            #                    is 0.
+
+        logger.info("Failure mode: %s", mode)
+        if mode == CONST.KILL_GCS_SERVER:
+            if new_head_pod_name != old_head_pod_name:
+                logger.warning(
+                    "GCS server process is killed. The head Pod should be restarted "
+                    "rather than creating a new head Pod. There is something wrong. "
+                    "new_head_pod_name: %s, old_head_pod_name: %s",
+                    new_head_pod_name, old_head_pod_name
+                )
+                return False
             if new_restart_count != old_restart_count + 1:
-                logger.info(f'new_restart_count != old_restart_count + 1 => new_restart_count: {new_restart_count},' +
-                    f' old_restart_count: {old_restart_count}')
+                logger.info(
+                    "new_restart_count != old_restart_count + 1 =>"
+                    "new_restart_count: %s; old_restart_count: %s",
+                    new_restart_count, old_restart_count
+                )
                 return False
-        # All pods should be "Running" and "Ready". This check is an overkill. We added this check due to
-        # the buggy behaviors of Ray HA. To elaborate, when GCS server is killed, the head pod should restart,
-        # but worker pods should not. However, currently, worker pods will also restart.
-        # See https://github.com/ray-project/kuberay/issues/634 for more details.
-        for pod in all_pods.items:
-            if pod.status.phase != 'Running':
-                logger.info(f'Pod {pod.metadata.name} is not Running.')
+        elif mode == CONST.KILL_HEAD_POD:
+            if new_head_pod_name == old_head_pod_name:
+                logger.info("The old head Pod %s is not killed.", old_head_pod_name)
+                return False
```

Comment:
Nit: do we need to include an `else` block and raise an error for unexpected mode?

Reply:
Updated https://github.com/ray-project/kuberay/pull/1341/commits/ead7ca3764a8aebaafe336e4ae6ecf991e23ac61